### ISSUE TITLE
threshold_ema_dead_code is in units of an EMA of per-batch counts, so a value copied from large-batch configs resamples most of the codebook every step at small batches

### ISSUE THREAD

**Issue by raunak2007** [lucidrains/vector-quantize-pytorch#257]
Follow-up to #256, which stopped `replace` from writing the same vector into several codes. Removing the duplicates turned out to change nothing downstream in the model I found them in, and the reason is the threshold itself, so I'm raising it as a design question rather than a PR.

**The unit problem.** `cluster_size` is an EMA (decay 0.8) of per-batch assignment counts, so its steady state for a code is roughly the number of vectors that hit it per batch. With `N` vectors per step and `K` codes that averages `N / K`. `threshold_ema_dead_code = 2` is the value the SoundStream / EnCodec configs use, where a batch holds thousands of frames; at `N / K < 2` most codes can never reach the threshold, and a freshly replaced code (reset to 2.0) needs at least two hits in the very next batch to survive (`0.8 * 2 + 0.2 * b >= 2`). So the same number means "expire codes that have gone quiet" at one batch size and "resample most of the codebook every step" at another, and nothing in the API says which regime you're in.

**Toy measurement** (`ResidualVQ(dim=128, num_quantizers=8, codebook_size=1024, kmeans_init=True, threshold_ema_dead_code=2, quantize_dropout=True)`, Gaussian inputs, 200 steps; codes sitting at the reset value after a step, i.e. replaced on that step):

```
 800 vectors/step (batch 16 x 50):  [1008, 882, 878, 1010, 872, 872, 867, 1011] of 1024 per layer
3200 vectors/step (batch 64 x 50):  [ 883, 850, 778,  895, 870, 883, 860,  867]
```

Gaussian data is the worst case (nothing clusters, so nothing accumulates counts), but the point is that at 800 vectors per step the codebook is not an EMA-learned codebook at all; it is a fresh sample of the batch.

**Real model.** A 9.8M-parameter SoundStream-style speech codec (8 x 1024 RVQ, 800 latent frames per batch), trained twice from scratch with the same seed and schedule on the post-#256 code, threshold 2 vs 0.2 (below the 0.78 mean count per code, so a code expires after roughly six consecutive zero-hit steps rather than every step):

| quantizer | perplexity, threshold 2 | perplexity, threshold 0.2 |
|---|---|---|
| 1 | 563 | 842 |
| 2 | 636 | 896 |
| 3 | 638 | 912 |
| 4 | 574 | 891 |
| 5 | 488 | 850 |
| 6 | 368 | 723 |
| 7 | 292 | 522 |
| 8 | 240 | 344 |

Log-mel distance on held-out speakers improved by 5 / 4 / 4 / 2 % at 0.5 / 1 / 2 / 4 kbps and SI-SNR by 0.8 / 0.5 / 0.3 / 0 dB, with the largest gain at the lowest bitrate, where the first codebook matters most. Repo with the runs and eval scripts: https://github.com/raunak2007/mini-codec.

**Two possible fixes**, and I'm happy to implement either:

1. Documentation only: state in the README and the `threshold_ema_dead_code` docstring that the threshold is compared against an EMA of per-batch counts whose scale is `vectors_per_step / codebook_size`, and that values from large-batch configs should be scaled down accordingly.
2. A warning at the first forward pass when `vectors_per_step / codebook_size < threshold_ema_dead_code`, saying that most codes will be expired every step, in the same spirit as the k-means init warning from #256. (An alternative would be a relative threshold, e.g. a fraction of the expected count per code, but that changes semantics and is more your call than mine.)

**Comment by raunak2007** [comment]
Adding the intelligibility axis for the two runs above (Whisper small.en WER on 500 held-out utterances, original audio 2.7%): threshold 2 gives 14.5 / 4.8 / 3.2 / 2.9 % at 0.5 / 1 / 2 / 4 kbps, threshold 0.2 gives 11.9 / 4.2 / 3.1 / 2.9 %. So the churn costs about 18% relative WER at the lowest bitrate, where the first codebook matters most.
